plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2546,6 +2552,62 @@ private static IcebergColumnHandle createProjectedColumnHandle(IcebergColumnHand
                 Optional.empty());
     }
 
+    @Override
+    public Optional<PartialSortApplicationResult<ConnectorTableHandle>> applyPartialSort(
+            ConnectorSession session,
+            ConnectorTableHandle tableHandle,
+            Map<ColumnHandle, io.trino.spi.connector.SortOrder> columnHandleSortOrderMap)
+    {
+        IcebergTableHandle originalTableHandle = (IcebergTableHandle) tableHandle;
+        if (originalTableHandle.isSort()) {
+            return Optional.empty();
+        }
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) tableHandle;
+        Table table;
+        table = catalog.loadTable(session, icebergTableHandle.getSchemaTableName());
+        if (table.sortOrder().isUnsorted()) {
+            return Optional.empty();
+        }
+        SortField sortField = table.sortOrder().fields().get(0);
+        IcebergColumnHandle columnHandle = (IcebergColumnHandle) columnHandleSortOrderMap.keySet().stream().toList().get(0);
+        if (!table.schema().findColumnName(sortField.sourceId()).equalsIgnoreCase(columnHandle.getName())) {
+            return Optional.empty();
+        }
+        io.trino.spi.connector.SortOrder trinoSortOrder = columnHandleSortOrderMap.values().stream().toList().get(0);
+        io.trino.spi.connector.SortOrder icebergSortOrder = toSortOrder(sortField);
+        boolean allSorted = true;
+        TupleDomain<IcebergColumnHandle> effectivePredicate =
+                icebergTableHandle.getUnenforcedPredicate().intersect(icebergTableHandle.getEnforcedPredicate());
+        try (CloseableIterable<FileScanTask> fileScanTasks = table.newScan().filter(toIcebergExpression(effectivePredicate)).planFiles()) {
+            for (FileScanTask scanTask : fileScanTasks) {
+                DataFile file = scanTask.file();
+                if (file.sortOrderId() == 0) {
+                    allSorted = false;
+                    break;
+                }
+            }
+        }
+        catch (Throwable t) {
+            throw new RuntimeException(t);
+        }
+        return Optional.of(
+                new PartialSortApplicationResult<>(
+                        originalTableHandle.withSort(allSorted),
+                        trinoSortOrder.isAscending() == icebergSortOrder.isAscending(),
+                        trinoSortOrder.isNullsFirst() == icebergSortOrder.isNullsFirst(),
+                        allSorted));
```

Comment:
I think i don't understand how it works.
Let's assume i have a bunch of sorted files.
Each of these files may produce one or more splits.
How does the fact that files are sorted help in anything?

Reply:
This is a good question that is worth exploring. If all the files in the query range are already sorted and the selectivity of the filter field is very low, then it is necessary to have one split per file. However, if the selectivity of the filter field is relatively high, one split per file may not have enough concurrency, which may affect query performance, because almost all the data in the file needs to be read to find the required data. In this case, we would rather have a file split into multiple splits to improve query performance.